### ISSUE TITLE
rust-analyzer uses an identically named method from a wrong trait

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/rust-analyzer#18526]
Using rust-analyzer on this [crate](https://github.com/RustCrypto/traits/tree/a54bd0d7fbe7ea3543ff36eafbfbbb1b309aaccb/aead) (note the commit hash) results in two false-positive compilation errors (i.e. compiling the same code with cargo produces no errors). One of the errors is on this [line](https://github.com/RustCrypto/traits/blob/a54bd0d7fbe7ea3543ff36eafbfbbb1b309aaccb/aead/src/lib.rs#L323):
```rust
    #[inline]
    fn encrypt_to_vec<'msg, 'aad>(
        &self,
        nonce: &Nonce<Self>,
        pt_payload: impl Into<Payload<'msg, 'aad>>,
    ) -> Result<Vec<u8>> {
        let mut buf = Vec::new();
        self.encrypt_to_buffer(nonce, pt_payload, &mut buf)?;
        // expected 4 arguments, found 3 (E0107)         ^
        Ok(buf)
    }
```

Clicking on the `encrypt_to_buffer` method jumps to the [`DynAead::encrypt_to_buffer`](https://github.com/RustCrypto/traits/blob/a54bd0d7fbe7ea3543ff36eafbfbbb1b309aaccb/aead/src/dyn_aead.rs#L58-L64) method, while [`Aead::encrypt_to_buffer`](https://github.com/RustCrypto/traits/blob/a54bd0d7fbe7ea3543ff36eafbfbbb1b309aaccb/aead/src/lib.rs#L356-L362) should've been used instead.

rust-analyzer version: 0.3.2188
rustc version: 1.84.0-nightly (5ec7d6eee 2024-11-17)

**Comment by lnicola** [comment]
Probably same as #5514.

**Comment by newpavlov** [comment]
Considering that the issue is 4 years old, it does not look like it. The code was working fine until I updated rust-analyzer today. Unfortunately, I don't know the exact version which I was using before.

**Comment by newpavlov** [comment]
v0.3.2180 handles the code without issues.

**Comment by vorporeal** [comment]
I started noticing the same thing in my codebase upon updating our Rust toolchain to 1.83 (from 1.82).

**Comment by Expurple** [comment]
I'm hitting the same problem on some (but not all) calls to [`sea_orm::EntityTrait::find()`](https://docs.rs/sea-orm/1.1.11/sea_orm/entity/trait.EntityTrait.html#method.find) in my proprietary codebase. `rust-analyzer` confuses this method with [`std::Iterator::find`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find) and reports a false-positive error: `expected 2 arguments, found 0 (E0107)`.

I use `rust-analyzer` in VSCode, currently at `0.3.2466`. I remember seeing this for a few months at least. Only now got to goodle/report it.

Ping me when the original reproducer stops reproducing the issue. I'll check on my project and make a new reproducer if the issue persists.

**Comment by lnicola** [comment]
The original code in RustCrypto seems to be gone, and we pick the right `clamp` at https://github.com/xiph/rav1e/blob/564ae3b0007ae2b06893fd7166bf88c5a84c5b63/src/predict.rs#L296.
